presto-main/src/main/java/io/prestosql/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -58,7 +61,36 @@
 
     public SymbolMapper(Map<Symbol, Symbol> mapping)
```

Comment:
Maybe we could have something more elaborate than `Map<Symbol, Symbol> mapping`. Some kind of entity that would maintain cycles prevention and short paths along the way? So the would be no map recreation. That could be then a permanent change, but surely more invasive change. WDYT?